tests/esrequest_test.php

```diff
@@ -33,25 +35,58 @@
 use \GuzzleHttp\HandlerStack;
 use \GuzzleHttp\Middleware;
 use \GuzzleHttp\Psr7\Response;
-use \GuzzleHttp\Psr7\Request;
 
 /**
  * Tests for esrequest class
  *
  * @package     search_elastic
  * @copyright   Matt Porritt <mattp@catalyst-au.net>
  * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * @covers      \search_elastic\esrequest
  */
-class search_elastic_esrequest_testcase extends advanced_testcase {
+class esrequest_test extends \advanced_testcase {
+
+    /**
+     * @var \stdClass $cfg Backup of global config.
+     */
+    protected $cfg;
+
+    /**
+     * Polyfill to support the new regexp assertion in place of the old, deprecated one.
+     *
+     * This can be removed once we no longer need to support Moodle <3.11/PHPUnit 8.5.
+     *
+     * @param string $method
+     * @param array $args
+     * @return mixed|void
+     */
+    public function __call(string $method, array $args) {
+        if ($method === 'assertMatchesRegularExpression' && method_exists($this, 'assertRegExp')) {
+            return call_user_func_array([$this, 'assertRegExp'], $args);
+        }
+    }
 
     /**
      * Test setup.
      */
     public function setUp(): void {
+        global $CFG;
+        $this->cfg = clone $CFG;
```

Comment:
why do we need this?

Reply:
If the development environment has proxy settings in config.php, this lets the test set them as required, then set them back to the original values in `tearDown()`. The previous approach of using `set_config()` wasn't reliable, as the code under test reads from `$CFG` directly so always got the values from config.php if they were set.

Reply:
So you're saying this change is required for a specific dev set up where you have some related variables set in config.php? If so, to me this doesn't sound as a good reason to change the code in the tests.
Can you please revert affected bits if they are not really impacting 4.1 upgrade and related to your local dev set up. Thanks!

Reply:
My thinking was that this makes the tests more robust, as currently they assume that `$CFG->proxy*` variables are unset. This method ensures the same behaviour of the test whether they are set in config.php or not. However if you'd rather not have this change here I am happy to remove it or separate it out.

If we do remove this I think it's worth at least skipping the tests with an explanation if the `proxy` variables are set in config.php, as they won't be testing the intended behaviour.

Reply:
Let's have a second opinion on this one.  @brendanheywood @danmarsden @Peterburnett ?

Reply:
I agree here for the most part with what Mark said. Our dev environtment does use a proxy, and this may be the case in the wild. These tests should ALWAYS pass irregardless of that. This is a bit of a hack, but does the job

Reply:
ok. Cool. 2 vs 1. You win. :)